### PR TITLE
#355 Added menard equations 2.21 and 2.22 of CUR 228

### PR DESCRIPTION
**Pull request by johan-tuls**
## Description

Added Menard equations CUR 228

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## Checklist:

- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            Factor dependent on soil type [-]:
+        """
+        super().__init__()
+        self.r = r
+        self.e_p = e_p
+        self.alpha = alpha
+
+    @staticmethod
+    def _evaluate(r: M, e_p: KPA, alpha: float) -> KN_M3:
+        """Return the Menard stiffness k_h when r >= 0.3 m [kN/m3]."""
```
Comment (PabloVasconez) [inline blueprints/codes/cur/cur_228/formula_2_22.py:40]:
```suggestion
        """Return the Menard stiffness k_h when r < 0.3 m [kN/m3]."""
```

File: tests/codes/cur/cur_228/test_formula_2_21.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+"""Test for formula 2.21 from CUR 228."""
+
+import pytest
+
+from blueprints.codes.cur.cur_228.formula_2_21 import Form2Dot21MenardStiffness
+
+
+class TestForm2Dot21MenardStiffness:
```
Comment (rickdegoeij) [inline tests/codes/cur/cur_228/test_formula_2_21.py]: add test with invalid values for e_p

File: blueprints/codes/cur/cur_228/formula_2_22.py (hunk 1 above)
Comment (rickdegoeij) [inline blueprints/codes/cur/cur_228/formula_2_22.py]: add checks for negative numbers
```suggestion
        """Return the Menard stiffness k_h when r >= 0.3 m [kN/m3]."""
        raise_if_negative(r=r, e_p=e_p, alpha=alpha)
```

File: tests/codes/cur/cur_228/test_formula_2_22.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
+"""Test for formula 2.22 from CUR 228."""
+
+import pytest
+
+from blueprints.codes.cur.cur_228.formula_2_22 import Form2Dot22MenardStiffness
+
+
+class TestForm2Dot22MenardStiffness:
```
Comment (rickdegoeij) [inline tests/codes/cur/cur_228/test_formula_2_22.py]: I would probably add some tests for the raise if negative checks. To make sure it's captured

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+from blueprints.type_alias import KN_M3, KPA, M
+
+
+class Form2Dot22MenardStiffness(Formula):
+    """Representation of equation 2.22 CUR 228."""
+
+    source_document = CUR_228
+    label = "2.22"
+
+    def __init__(self, r: M, e_p: KPA, alpha: float) -> None:
```
Comment (rickdegoeij) [inline blueprints/codes/cur/cur_228/formula_2_22.py]: Use dimensionless instead of float

```suggestion
from blueprints.type_alias import KN_M3, KPA, M, DIMENSIONLESS


class Form2Dot22MenardStiffness(Formula):
    """Representation of equation 2.22 CUR 228."""

    source_document = CUR_228
    label = "2.22"

    def __init__(self, r: M, e_p: KPA, alpha: DIMENSIONLESS) -> None:
```

File: blueprints/codes/cur/cur_228/formula_2_21.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+
+class Form2Dot21MenardStiffness(Formula):
+    """Representation of equation 2.21 CUR 228."""
+
+    source_document = CUR_228
+    label = "2.21"
+    r_0: M = 0.3
+
+    def __init__(self, r: M, e_p: KPA, alpha: float) -> None:
```
Comment (rickdegoeij) [inline blueprints/codes/cur/cur_228/formula_2_21.py]: Use DIMENSIONLESS instead of float
```suggestion
from blueprints.type_alias import KN_M3, KPA, M, DIMENSIONLESS
from blueprints.validations import raise_if_negative


class Form2Dot21MenardStiffness(Formula):
    """Representation of equation 2.21 CUR 228."""

    source_document = CUR_228
    label = "2.21"
    r_0: M = 0.3

    def __init__(self, r: M, e_p: KPA, alpha: DIMENSIONLESS) -> None:
```

**Review by rickdegoeij — CHANGES_REQUESTED** [review]
bit nitpicky sometimes, so feel free to interpret the comments :)

Comments that need address:
 - Latex equations don't match implementation
 - unittests for negative values
 - add negative value checks in 2.22
 - unittest for correctness of latex equations

**Review by egarciamendez — CHANGES_REQUESTED** [review]
check comments from @rickdegoeij

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        """
+        super().__init__()
+        self.r = r
+        self.e_p = e_p
+        self.alpha = alpha
+
+    @staticmethod
+    def _evaluate(r: M, e_p: KPA, alpha: float) -> KN_M3:
+        """Return the Menard stiffness k_h when r >= 0.3 m [kN/m3]."""
+        if r < 0.3:
```
[the remark this answers is not in the document]
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_22.py:41]: Set to cur_228 init file.

File: blueprints/codes/cur/cur_228/formula_2_21.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        -------
+        LatexFormula
+            Latex representation of the equation
+
+        """
+        n = n_decimals
+
+        return LatexFormula(
+            return_symbol="k_{h}",
+            equation=r"\frac{1}{3 \cdot E_{p}} \cdot "
```
[the remark this answers is not in the document]
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_21.py:69]: True, however the formula is ment to calculate k_H, in the document 1/k_h has been chosen for readability. So I will stick with k_h

File: tests/codes/cur/cur_228/test_formula_2_21.py (hunk 2 above)
Comment (johan-tuls) [reply tests/codes/cur/cur_228/test_formula_2_21.py]: Added

File: blueprints/codes/cur/cur_228/formula_2_22.py (hunk 1 above)
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_22.py]: Added

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            return e_p / 2 / r / ((4 * 2.65**alpha + 3 * alpha) / 18)
+        msg = "Radius is equal to- or larger than 0.3m, use: Eq2Dot21MenardStiffness"
```
[the remark this answers is not in the document]
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_22.py]: Constant set

File: blueprints/codes/cur/cur_228/__init__.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+"""CUR-228 package."""
+
+from blueprints.type_alias import M
+
+CUR_228 = "CUR 228"
+
+R_0: M = 0.3
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/__init__.py:7]: Why this R_0? Can you add some comments?

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    n_decimals: int = 2
+
+    def __init__(self, r: M, e_p: KPA, alpha: DIMENSIONLESS) -> None:
+        """Calculates the modulus of horizontal subgrade reaction (k_h) using Menard stiffness for r < 0.3 m.
+
+        Parameters
+        ----------
+        r: M
+            The radius of a foundation pile [m]:
+            r >= 0.3 m
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/formula_2_22.py:24]: < 0.3

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        """
+        super().__init__()
+        self.r = r
+        self.e_p = e_p
+        self.alpha = alpha
+
+    @staticmethod
+    def _evaluate(r: M, e_p: KPA, alpha: DIMENSIONLESS) -> KN_M3:
+        """Return the Menard stiffness k_h when r < 0.3 m [kN/m3]."""
+        raise_if_less_or_equal_to_zero(e_p=e_p, r=r, alpha=alpha)
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/formula_2_22.py:43]: check the raise_if_neg vs raise_if_less_or_equal dif here too.

File: blueprints/type_alias.py [hunk 12]
```diff
@@ -95,6 +95,8 @@
 """Newton per square meter (N/m²), represented as a float."""
 KN_M2 = float
 """Kilonewton per square meter (kN/m²), represented as a float."""
+KN_M3 = float
+"""Kilonewton per cubic meter (kN/m²), represented as a float."""
```
Comment (GerjanDorgelo) [inline blueprints/type_alias.py:99]: there is a ^2 in this comment. should be ^3.

File: docs/source/codes/cur/cur_228/formulas.md [hunk 13]
```diff
@@ -0,0 +1,12 @@
+**CUR 228 - 2010
+Design guideline for soil laterally loaded piles**
+
+The table presents a list of formulas from the CUR 228 guideline, tracking their implementation status
+( :x: or :heavy_check_mark: ) and any pertinent remarks. The 'Object Name' column references the corresponding Python entities inside of Blueprints.
+
+Total of 2 formulas present.
```
Comment (GerjanDorgelo) [inline docs/source/codes/cur/cur_228/formulas.md:7]: this one is incomplete. could you add that as a note?

File: pyproject.toml [hunk 14]
```diff
@@ -54,3 +54,5 @@ python_version = "3.12"
 [[tool.mypy.overrides]]
 module = ["matplotlib.*"]
 ignore_missing_imports = true
+
```
Comment (GerjanDorgelo) [inline pyproject.toml:57]: these changes can be reverted

File: tests/codes/cur/cur_228/test_formula_2_21.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+    def test_latex_method(self) -> None:
+        """Test the latex method."""
+        r = 0.5  # m
+        e_p = 2.47  # kN/m²
+        alpha = 1 / 3  # -
+        form_2_21 = Form2Dot21ModulusHorizontalSubgrade(r=r, e_p=e_p, alpha=alpha)
+
+        # Test the full LaTeX representation
+        assert isinstance(form_2_21.latex().complete, str)
```
Comment (GerjanDorgelo) [inline tests/codes/cur/cur_228/test_formula_2_21.py:57]: could you also test the actual response of the function? see prompt template files.

File: tests/codes/cur/cur_228/test_formula_2_22.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+    def test_latex_method(self) -> None:
+        """Test the latex method."""
+        r = 0.2  # m
+        e_p = 2.47  # kN/m²
+        alpha = 1 / 3  # -
+        form_2_22 = Form2Dot22ModulusHorizontalSubgrade(r=r, e_p=e_p, alpha=alpha)
+
+        # Test the full LaTeX representation
+        assert isinstance(form_2_22.latex().complete, str)
```
Comment (GerjanDorgelo) [inline tests/codes/cur/cur_228/test_formula_2_22.py:55]: could you also test the actual response of the function? see prompt template files.

File: tests/codes/cur/cur_228/test_formula_2_21.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+        # Expected result, manually calculated
+        manually_calculated_result = 9.187357198
+
+        assert form_2_21 == pytest.approx(expected=manually_calculated_result, rel=1e-4)
+
+    @pytest.mark.parametrize(
+        ("e_p", "r", "alpha"),
+        [
+            (-500.0, -0.6, -0.33),
```
Comment (GerjanDorgelo) [inline tests/codes/cur/cur_228/test_formula_2_21.py:28]: can you test these individually? now it crashes on -500 being negative and not per se on -0.6 being negative.

File: tests/codes/cur/cur_228/test_formula_2_22.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        e_p = 2.47  # kN/m²
+        alpha = 1 / 3  # -
+
+        with pytest.raises(ValueError):
+            Form2Dot22ModulusHorizontalSubgrade(r=r, e_p=e_p, alpha=alpha)
+
+    @pytest.mark.parametrize(
+        ("e_p", "r", "alpha"),
+        [
+            (-500.0, -0.2, -0.33),
```
Comment (GerjanDorgelo) [inline tests/codes/cur/cur_228/test_formula_2_22.py:38]: can you test these individually too?

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        Parameters
+        ----------
+        r: M
+            The radius of a foundation pile [m]:
+            r >= 0.3 m
+        e_p: KPA
+            Elastic modulus of Ménard [kPa]:
+            e_p ≈ beta * q_c
+                beta: DIMENSIONLESS
+                    Dependent on soil type [-]:
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/formula_2_22.py:29]: cna you add reference to table 2.1?

File: blueprints/codes/cur/cur_228/formula_2_22.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+                q_c: KPA
+                    Cone resistance [kPa]
+        alpha: DIMENSIONLESS
+            Factor dependent on soil type [-]:
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/formula_2_22.py:33]: cna you add reference to table 2.1?

File: blueprints/codes/cur/cur_228/formula_2_21.py [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+        Parameters
+        ----------
+        r: M
+            The radius of a foundation pile [m]:
+            r >= 0.3 m
+        e_p: KPA
+            Elastic modulus of Ménard [kPa]:
+            e_p ≈ beta * q_c
+            beta: DIMENSIONLESS
+                Dependent on soil type [-]:
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/formula_2_21.py:29]: cna you add reference to table 2.1?

File: blueprints/codes/cur/cur_228/formula_2_21.py [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+            q_c: KPA
+                Cone resistance [kPa]
+        alpha: DIMENSIONLESS
+            Factor dependent on soil type [-]:
```
Comment (GerjanDorgelo) [inline blueprints/codes/cur/cur_228/formula_2_21.py:33]: cna you add reference to table 2.1?

File: blueprints/codes/cur/cur_228/__init__.py (hunk 9 above)
Comment (GerjanDorgelo) [reply blueprints/codes/cur/cur_228/__init__.py:7]: After review: maybe add this to the two relevant files itself. Not as a general parameter.

File: blueprints/codes/cur/__init__.py [hunk 23]
```diff
@@ -0,0 +1 @@
+"""Cur guidelines."""
```
[the remark this answers is not in the document]
Comment (johan-tuls) [reply blueprints/codes/cur/__init__.py:1]: Done, but see PR 520...

File: blueprints/codes/cur/cur_228/__init__.py (hunk 9 above)
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/__init__.py:7]: Yeah it is part of the document at constants. Instead of a constant file inside the cur_228 i thought add it to the init. As it is used in multiple files.

File: blueprints/codes/cur/cur_228/formula_2_22.py (hunk 10 above)
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_22.py:24]: Fixed

File: blueprints/codes/cur/cur_228/formula_2_22.py (hunk 11 above)
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_22.py:43]: r and alpha cannot be 0. e_p could be 0 if qc is 0. So changed that. Idem for other item

File: blueprints/codes/cur/cur_228/formula_2_21.py (hunk 21 above)
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_21.py:29]: Added

File: blueprints/codes/cur/cur_228/formula_2_21.py (hunk 22 above)
Comment (johan-tuls) [reply blueprints/codes/cur/cur_228/formula_2_21.py:33]: Added

File: blueprints/type_alias.py (hunk 12 above)
Comment (johan-tuls) [reply blueprints/type_alias.py:99]: changed

File: pyproject.toml (hunk 14 above)
Comment (johan-tuls) [reply pyproject.toml:57]: white line removed

File: tests/codes/cur/cur_228/test_formula_2_21.py (hunk 15 above)
Comment (johan-tuls) [reply tests/codes/cur/cur_228/test_formula_2_21.py:57]: done

File: tests/codes/cur/cur_228/test_formula_2_21.py (hunk 17 above)
Comment (johan-tuls) [reply tests/codes/cur/cur_228/test_formula_2_21.py:28]: Done

**Comment by johan-tuls** [comment]
@GerjanDorgelo As the latex formula was also used in #522 please review the changes in that PR. All commented have been fixed.

**Comment by GerjanDorgelo** [comment]
> @GerjanDorgelo As the latex formula was also used in #522 please review the changes in that PR. All commented have been fixed.

@johan-tuls i dont see any updates. Any chance you forgot to sync the changes?

**Comment by GerjanDorgelo** [comment]
i see this branch has been merged into #522. lets close this one then.
